src/useScrollbarSize.ts

```diff
@@ -0,0 +1,59 @@
+import { useEffect, useRef, useState } from 'react';
+import { debounce } from './utils';
+
+export interface ScrollbarMeasurements {
+	/** Current height of the scrollbar */
+	height: number;
+	/** Current width of the scrollbar */
+	width: number;
+}
+
+const useScrollbarSize = (): ScrollbarMeasurements => {
+	const [height, setHeight] = useState(0);
+	const [width, setWidth] = useState(0);
+	const element = useRef<HTMLDivElement | null>(null);
+
+	// initialize resize event handler and state when mounted
+	useEffect(() => {
+		const getElement = () => {
+			if (element.current == null) {
+				// element was not created yet -- initialize
+				element.current = document.createElement('div');
+				element.current.style.width = '99px';
+				element.current.style.height = '99px';
+				element.current.style.overflow = 'scroll';
+				element.current.style.position = 'absolute';
+				element.current.style.top = '-9999px';
+				element.current.setAttribute('aria-hidden', 'true');
+				element.current.setAttribute('role', 'presentation');
+			}
+			return element.current;
+		};
+
+		const updateState = () => {
+			const { offsetHeight, clientHeight, offsetWidth, clientWidth } = getElement();
+			const scrollbarHeight = offsetHeight - clientHeight;
+			const scrollbarWidth = offsetWidth - clientWidth;
+			setHeight((h) => (h !== scrollbarHeight ? scrollbarHeight : h));
+			setWidth((w) => (w !== scrollbarWidth ? scrollbarWidth : w));
```

Comment:
question: doing setHeight and setWidth causes two renders if react decides not to batch it right? react should batch it right? but is there chance it doesn't?

if there is chance it doesnt batch, then what about this thought:

const [dimensions, setDimensions] = useState({ height: 0, width: 0} );

I proposed this because it seems we always set height/width together and at the bottom it returns height and width in an object, which is referenitally new every time this hook runs. With the above we now get new object only when dimensions change.

We can check in here against old/new height/width if we should do a setDimensions to ensure we don't create new objects on reszie if not necessary.

Reply:
> question: doing setHeight and setWidth causes two renders if react decides not to batch it right? react should batch it right? but is there chance it doesn't?

See: https://github.com/facebook/react/issues/10231#issuecomment-316644950

Apart from the initial call, these will be batched because they're happening in the context of an event handler (resize).

>I proposed this because it seems we always set height/width together and at the bottom it returns height and width in an object, which is referenitally new every time this hook runs. With the above we now get new object only when dimensions change.

This is fair.  👍

Reply:
Oh thanks for sharing that comment! That is awesome that they batch as much as possible by default!

Reply:
Thanks for the feedback here, I've made the changes as discussed.